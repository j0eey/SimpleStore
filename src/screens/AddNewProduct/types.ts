import { Asset } from "react-native-image-picker";
import { LocationType } from "../../types/types";

export interface FormHeaderProps {
  title: string;
}

export interface ImageUploadSectionProps {
  images: Asset[];
  onImagePick: () => void;
  onCameraCapture: () => void;
  onRemoveImage: (index: number) => void;
  isSubmitting: boolean;
}

export interface FormInputProps {
  label: string;
  value: string;
  onChangeText: (text: string) => void;
  placeholder: string;
  multiline?: boolean;
  numberOfLines?: number;
  keyboardType?: 'default' | 'numeric';
  isSubmitting: boolean;
}

export interface LocationSelectorProps {
  location: LocationType;
  onLocationSelect: () => void;
  isSubmitting: boolean;
}

export interface FormActionsProps {
  onSubmit: () => void;
  onCancel: () => void;
  isSubmitting: boolean;
}