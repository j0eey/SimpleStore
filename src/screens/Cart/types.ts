import { Animated } from "react-native";
import { CartItemType } from "../../types/Cart";

export interface CartHeaderProps {
  itemCount: number;
}

export interface EmptyCartProps {
  onStartShopping: () => void;
}

export interface CartItemProps {
  item: CartItemType;
  onIncreaseQuantity: (itemId: string, currentQuantity: number) => void;
  onDecreaseQuantity: (itemId: string, currentQuantity: number) => void;
  onRemoveItem: (itemId: string) => void;
}

export interface CartSummaryProps {
  cart: CartItemType[];
  summaryExpanded: boolean;
  summaryHeight: Animated.Value;
  onToggleSummary: () => void;
  onCheckout: () => void;
  onClearCart: () => void;
  calculateTotal: () => string;
}

export interface CartListProps {
  cart: CartItemType[];
  onIncreaseQuantity: (itemId: string, currentQuantity: number) => void;
  onDecreaseQuantity: (itemId: string, currentQuantity: number) => void;
  onRemoveItem: (itemId: string) => void;
}