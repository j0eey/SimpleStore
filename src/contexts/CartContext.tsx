import React, { createContext, useState, useContext, ReactNode, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { CartItem, CartContextType } from '../types/Cart';
import { Product } from '../types/Product';
import Toast from 'react-native-toast-message';


const CART_STORAGE_KEY = '@myMarket_cart';

const CartContext = createContext<CartContextType | undefined>(undefined);

interface CartProviderProps {
  children: ReactNode;
}

export const CartProvider = ({ children }: CartProviderProps) => {
  const [cart, setCart] = useState<CartItem[]>([]);
  const [isLoaded, setIsLoaded] = useState(false); 

  useEffect(() => {
    const loadCart = async () => {
      try {
        const storedCart = await AsyncStorage.getItem(CART_STORAGE_KEY);
        if (storedCart) {
          setCart(JSON.parse(storedCart));
        }
      } catch (error) {
        Toast.show({
          type: 'error',
          text1: 'Failed to load shopping cart.',
          text2: 'Please restart the app if issues persist.'
        });
      } finally {
        setIsLoaded(true);
      }
    };
    loadCart();
  }, []);

  useEffect(() => {
    const saveCart = async () => {
      try {
        await AsyncStorage.setItem(CART_STORAGE_KEY, JSON.stringify(cart));
      } catch (error) {
      }
    };
    if (isLoaded) {
      saveCart();
    }
  }, [cart, isLoaded]);

  const addToCart = (product: Product, quantityToAdd: number = 1) => {
    setCart(prevCart => {
      const existingItemIndex = prevCart.findIndex(item => item._id === product._id);

      if (existingItemIndex > -1) {
        const updatedCart = [...prevCart];
        const currentQuantity = updatedCart[existingItemIndex].quantity;
        const newQuantity = currentQuantity + 1;
        updatedCart[existingItemIndex] = {
          ...updatedCart[existingItemIndex],
          quantity: newQuantity,
        };
        Toast.show({
          type: 'info',
          text1: `Quantity for "${product.title}" updated!`,
          text2: `New quantity: ${newQuantity}`,
        });
        return updatedCart;
      } else {
        const newItem: CartItem = { ...product, quantity: quantityToAdd };
        Toast.show({
          type: 'success',
          text1: `"${product.title}" added to cart!`,
          text2: `Quantity: ${quantityToAdd}`,
        });
        return [...prevCart, newItem];
      }
    });
  };

  const removeFromCart = (productId: string) => {
    setCart(prevCart => {
      const updatedCart = prevCart.filter(item => item._id !== productId);
      Toast.show({
        type: 'success',
        text1: 'Item removed from cart.',
      });
      return updatedCart;
    });
  };

  const updateCartItemQuantity = (productId: string, newQuantity: number) => {
    setCart(prevCart => {
      if (newQuantity <= 0) {
        return prevCart.filter(item => item._id !== productId);
      }

      const updatedCart = prevCart.map(item =>
        item._id === productId ? { ...item, quantity: newQuantity } : item
      );
      return updatedCart;
    });
  };

  const clearCart = () => {
    setCart([]);
    Toast.show({
      type: 'info',
      text1: 'Cart cleared!',
    });
  };

  return (
    <CartContext.Provider value={{ cart, addToCart, removeFromCart, updateCartItemQuantity, clearCart }}>
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error('useCart must be used within a CartProvider');
  }
  return context;
};
