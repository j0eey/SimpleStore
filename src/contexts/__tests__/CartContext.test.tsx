import React from 'react';
import { render, act, waitFor } from '@testing-library/react-native';
import { Text } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Toast from 'react-native-toast-message';
import { CartProvider, useCart } from '../CartContext';
import { Product } from '../../types/Product';
import { CartItem } from '../../types/Cart';

// AsyncStorage and Toast are already mocked in jest.setup.js
// No need to mock them again here

// Test component to access cart context
const TestComponent = ({ onRender }: { onRender?: (cartData: any) => void }) => {
  const cartContext = useCart();
  
  React.useEffect(() => {
    if (onRender) {
      onRender(cartContext);
    }
  }, [cartContext, onRender]);

  return (
    <>
      <Text testID="cart-count">{cartContext.cart.length}</Text>
      <Text testID="cart-items">{JSON.stringify(cartContext.cart)}</Text>
    </>
  );
};

// Test component for error boundary testing
const TestComponentWithoutProvider = () => {
  try {
    useCart();
    return <Text testID="no-error">No Error</Text>;
  } catch (error) {
    return <Text testID="error">{(error as Error).message}</Text>;
  }
};

describe('CartContext', () => {
  const mockProduct: Product = {
    _id: 'test-product-1',
    title: 'Test Product',
    price: 29.99,
    description: 'A test product',
    images: [
      {
        url: 'test-image.jpg',
        _id: 'img-1',
        fullUrl: 'https://example.com/test-image.jpg',
      }
    ],
    location: {
      name: 'Test Location',
      latitude: 40.7128,
      longitude: -74.0060,
    },
    user: {
      _id: 'user-1',
      email: 'test@example.com',
    },
    createdAt: '2024-01-01T00:00:00Z',
    updatedAt: '2024-01-01T00:00:00Z',
    __v: 0,
  };

  const mockProduct2: Product = {
    _id: 'test-product-2',
    title: 'Test Product 2',
    price: 49.99,
    description: 'Another test product',
    images: [
      {
        url: 'test-image-2.jpg',
        _id: 'img-2',
        fullUrl: 'https://example.com/test-image-2.jpg',
      }
    ],
    location: {
      name: 'Test Location 2',
      latitude: 41.8781,
      longitude: -87.6298,
    },
    user: {
      _id: 'user-2',
      email: 'test2@example.com',
    },
    createdAt: '2024-01-02T00:00:00Z',
    updatedAt: '2024-01-02T00:00:00Z',
    __v: 0,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    AsyncStorage.clear();
  });

  // SNAPSHOT TESTS - for cart state structures
  describe('Cart State Snapshots', () => {
    it('should match snapshot for initial empty cart state', async () => {
      let cartData: any;
      
      const TestWrapper = () => (
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      render(<TestWrapper />);

      await waitFor(() => {
        expect(cartData).toBeDefined();
        expect(cartData.cart).toBeDefined();
      });

      expect({
        cart: cartData.cart,
        cartLength: cartData.cart.length,
        availableMethods: Object.keys(cartData).filter(key => typeof cartData[key] === 'function'),
      }).toMatchSnapshot('initial-empty-cart-state');
    });

    it('should match snapshot for cart with single item', async () => {
      let cartData: any;
      
      const TestWrapper = () => (
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      render(<TestWrapper />);

      await waitFor(() => {
        expect(cartData).toBeDefined();
      });

      // Add item to cart
      act(() => {
        cartData.addToCart(mockProduct, 1);
      });

      await waitFor(() => {
        expect(cartData.cart.length).toBe(1);
      });

      expect({
        cart: cartData.cart,
        cartLength: cartData.cart.length,
        firstItem: cartData.cart[0],
      }).toMatchSnapshot('cart-with-single-item');
    });

    it('should match snapshot for cart with multiple items', async () => {
      let cartData: any;
      
      const TestWrapper = () => (
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      render(<TestWrapper />);

      await waitFor(() => {
        expect(cartData).toBeDefined();
      });

      // Add multiple items
      act(() => {
        cartData.addToCart(mockProduct, 2);
        cartData.addToCart(mockProduct2, 1);
      });

      await waitFor(() => {
        expect(cartData.cart.length).toBe(2);
      });

      expect({
        cart: cartData.cart,
        cartLength: cartData.cart.length,
        totalItems: cartData.cart.reduce((sum: number, item: CartItem) => sum + item.quantity, 0),
        productIds: cartData.cart.map((item: CartItem) => item._id),
      }).toMatchSnapshot('cart-with-multiple-items');
    });

    it('should match snapshot for cart operations sequence', async () => {
      let cartData: any;
      const operations: any[] = [];
      
      const TestWrapper = () => (
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      render(<TestWrapper />);

      await waitFor(() => {
        expect(cartData).toBeDefined();
      });

      // Sequence of operations
      act(() => {
        cartData.addToCart(mockProduct, 1);
      });
      await waitFor(() => expect(cartData.cart.length).toBe(1));
      operations.push({ action: 'add', cart: [...cartData.cart] });

      act(() => {
        cartData.addToCart(mockProduct, 1); // Increase quantity
      });
      await waitFor(() => expect(cartData.cart[0].quantity).toBe(2));
      operations.push({ action: 'add_existing', cart: [...cartData.cart] });

      act(() => {
        cartData.updateCartItemQuantity(mockProduct._id, 5);
      });
      await waitFor(() => expect(cartData.cart[0].quantity).toBe(5));
      operations.push({ action: 'update_quantity', cart: [...cartData.cart] });

      act(() => {
        cartData.addToCart(mockProduct2, 2);
      });
      await waitFor(() => expect(cartData.cart.length).toBe(2));
      operations.push({ action: 'add_second_product', cart: [...cartData.cart] });

      act(() => {
        cartData.removeFromCart(mockProduct._id);
      });
      await waitFor(() => expect(cartData.cart.length).toBe(1));
      operations.push({ action: 'remove_first_product', cart: [...cartData.cart] });

      expect(operations).toMatchSnapshot('cart-operations-sequence');
    });

    it('should match snapshot for quantity update scenarios', async () => {
      let cartData: any;
      
      const TestWrapper = () => (
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      render(<TestWrapper />);

      await waitFor(() => {
        expect(cartData).toBeDefined();
      });

      // Add item first
      act(() => {
        cartData.addToCart(mockProduct, 3);
      });

      await waitFor(() => {
        expect(cartData.cart.length).toBe(1);
      });

      const quantityScenarios = [
        { newQuantity: 5, description: 'increase quantity' },
        { newQuantity: 2, description: 'decrease quantity' },
        { newQuantity: 1, description: 'set to minimum' },
        { newQuantity: 0, description: 'set to zero (should remove)' },
      ];

      const results = [];
      for (const scenario of quantityScenarios) {
        act(() => {
          cartData.updateCartItemQuantity(mockProduct._id, scenario.newQuantity);
        });
        
        await waitFor(() => {
          // Wait for state to update
          if (scenario.newQuantity <= 0) {
            expect(cartData.cart.length).toBe(0);
          } else {
            expect(cartData.cart.length).toBe(1);
          }
        });
        
        results.push({
          scenario: scenario.description,
          newQuantity: scenario.newQuantity,
          cartState: [...cartData.cart],
        });

        // Re-add product for next scenario if it was removed
        if (scenario.newQuantity <= 0) {
          act(() => {
            cartData.addToCart(mockProduct, 3);
          });
          await waitFor(() => expect(cartData.cart.length).toBe(1));
        }
      }

      expect(results).toMatchSnapshot('quantity-update-scenarios');
    });
  });

  describe('Toast Notification Snapshots', () => {
    it('should match snapshot for all toast notification calls', async () => {
      let cartData: any;
      
      const TestWrapper = () => (
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      render(<TestWrapper />);

      await waitFor(() => {
        expect(cartData).toBeDefined();
      });

      // Clear previous calls
      (Toast.show as jest.Mock).mockClear();

      // Perform various operations to trigger toasts
      act(() => {
        cartData.addToCart(mockProduct, 1); // New item toast
        cartData.addToCart(mockProduct, 1); // Update quantity toast
        cartData.addToCart(mockProduct2, 2); // Another new item toast
      });

      await waitFor(() => {
        expect(cartData.cart.length).toBe(2);
      });

      act(() => {
        cartData.removeFromCart(mockProduct._id); // Remove item toast
        cartData.clearCart(); // Clear cart toast
      });

      await waitFor(() => {
        expect(cartData.cart.length).toBe(0);
      });

      const toastCalls = (Toast.show as jest.Mock).mock.calls;
      expect(toastCalls).toMatchSnapshot('toast-notification-calls');
    });
  });

  // UNIT TESTS - for behavior and functionality
  describe('CartProvider Behavior', () => {
    it('should provide initial empty cart', async () => {
      let cartData: any;
      
      render(
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      await waitFor(() => {
        expect(cartData).toBeDefined();
        expect(cartData.cart).toEqual([]);
      });
    });

    it('should load cart from AsyncStorage on mount', async () => {
      const storedCart: CartItem[] = [
        { ...mockProduct, quantity: 2 },
        { ...mockProduct2, quantity: 1 },
      ];

      (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify(storedCart));

      let cartData: any;
      
      render(
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      await waitFor(() => {
        expect(cartData).toBeDefined();
        expect(cartData.cart).toEqual(storedCart);
      });
    });

    it('should handle AsyncStorage loading errors gracefully', async () => {
      (AsyncStorage.getItem as jest.Mock).mockRejectedValue(new Error('Storage error'));

      let cartData: any;
      
      render(
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      await waitFor(() => {
        expect(cartData).toBeDefined();
        expect(cartData.cart).toEqual([]);
      });

      expect(Toast.show).toHaveBeenCalledWith({
        type: 'error',
        text1: 'Failed to load shopping cart.',
        text2: 'Please restart the app if issues persist.'
      });
    });

    it('should save cart to AsyncStorage when cart changes', async () => {
      let cartData: any;
      
      render(
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      await waitFor(() => {
        expect(cartData).toBeDefined();
      });

      act(() => {
        cartData.addToCart(mockProduct, 1);
      });

      await waitFor(() => {
        expect(AsyncStorage.setItem).toHaveBeenCalledWith(
          '@myMarket_cart',
          JSON.stringify([{ ...mockProduct, quantity: 1 }])
        );
      });
    });
  });

  describe('Cart Operations', () => {
    let cartData: any;

    beforeEach(async () => {
      render(
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      await waitFor(() => {
        expect(cartData).toBeDefined();
      });
    });

    describe('addToCart', () => {
      it('should add new item to cart', async () => {
        act(() => {
          cartData.addToCart(mockProduct, 1);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(1);
          expect(cartData.cart[0]).toEqual({ ...mockProduct, quantity: 1 });
        });
        
        expect(Toast.show).toHaveBeenCalledWith({
          type: 'success',
          text1: `"${mockProduct.title}" added to cart!`,
          text2: 'Quantity: 1',
        });
      });

      it('should increase quantity for existing item', async () => {
        act(() => {
          cartData.addToCart(mockProduct, 1);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(1);
        });

        act(() => {
          cartData.addToCart(mockProduct, 1);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(1);
          expect(cartData.cart[0].quantity).toBe(2);
        });
        
        expect(Toast.show).toHaveBeenCalledWith({
          type: 'info',
          text1: `Quantity for "${mockProduct.title}" updated!`,
          text2: 'New quantity: 2',
        });
      });

      it('should add multiple different products', async () => {
        act(() => {
          cartData.addToCart(mockProduct, 2);
          cartData.addToCart(mockProduct2, 1);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(2);
          expect(cartData.cart[0]).toEqual({ ...mockProduct, quantity: 2 });
          expect(cartData.cart[1]).toEqual({ ...mockProduct2, quantity: 1 });
        });
      });

      it('should use default quantity of 1 when not specified', async () => {
        act(() => {
          cartData.addToCart(mockProduct);
        });

        await waitFor(() => {
          expect(cartData.cart[0].quantity).toBe(1);
        });
      });
    });

    describe('removeFromCart', () => {
      it('should remove item from cart', async () => {
        act(() => {
          cartData.addToCart(mockProduct, 1);
          cartData.addToCart(mockProduct2, 1);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(2);
        });

        act(() => {
          cartData.removeFromCart(mockProduct._id);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(1);
          expect(cartData.cart[0]._id).toBe(mockProduct2._id);
        });
        
        expect(Toast.show).toHaveBeenCalledWith({
          type: 'success',
          text1: 'Item removed from cart.',
        });
      });

      it('should handle removing non-existent item gracefully', async () => {
        act(() => {
          cartData.addToCart(mockProduct, 1);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(1);
        });

        act(() => {
          cartData.removeFromCart('non-existent-id');
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(1);
          expect(cartData.cart[0]._id).toBe(mockProduct._id);
        });
      });
    });

    describe('updateCartItemQuantity', () => {
      beforeEach(async () => {
        act(() => {
          cartData.addToCart(mockProduct, 3);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(1);
        });
      });

      it('should update item quantity', async () => {
        act(() => {
          cartData.updateCartItemQuantity(mockProduct._id, 5);
        });

        await waitFor(() => {
          expect(cartData.cart[0].quantity).toBe(5);
        });
      });

      it('should remove item when quantity is set to 0', async () => {
        act(() => {
          cartData.updateCartItemQuantity(mockProduct._id, 0);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(0);
        });
      });

      it('should remove item when quantity is negative', async () => {
        act(() => {
          cartData.updateCartItemQuantity(mockProduct._id, -1);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(0);
        });
      });

      it('should handle updating non-existent item gracefully', async () => {
        act(() => {
          cartData.updateCartItemQuantity('non-existent-id', 5);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(1);
          expect(cartData.cart[0].quantity).toBe(3); // Original quantity unchanged
        });
      });
    });

    describe('clearCart', () => {
      it('should clear all items from cart', async () => {
        act(() => {
          cartData.addToCart(mockProduct, 1);
          cartData.addToCart(mockProduct2, 2);
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(2);
        });

        act(() => {
          cartData.clearCart();
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(0);
        });
        
        expect(Toast.show).toHaveBeenCalledWith({
          type: 'info',
          text1: 'Cart cleared!',
        });
      });

      it('should handle clearing empty cart', async () => {
        act(() => {
          cartData.clearCart();
        });

        await waitFor(() => {
          expect(cartData.cart).toHaveLength(0);
        });
      });
    });
  });

  describe('useCart Hook', () => {
    it('should throw error when used outside CartProvider', () => {
      const { getByTestId } = render(<TestComponentWithoutProvider />);
      
      expect(getByTestId('error')).toBeTruthy();
      expect(getByTestId('error').props.children).toBe('useCart must be used within a CartProvider');
    });

    it('should provide cart context when used within CartProvider', async () => {
      let cartData: any;
      
      render(
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      await waitFor(() => {
        expect(cartData).toBeDefined();
        expect(cartData.cart).toBeDefined();
        expect(typeof cartData.addToCart).toBe('function');
        expect(typeof cartData.removeFromCart).toBe('function');
        expect(typeof cartData.updateCartItemQuantity).toBe('function');
        expect(typeof cartData.clearCart).toBe('function');
      });
    });
  });

  describe('AsyncStorage Integration', () => {
    it('should not save to storage before initial load completes', async () => {
      (AsyncStorage.setItem as jest.Mock).mockClear();
      
      render(
        <CartProvider>
          <TestComponent />
        </CartProvider>
      );

      // Should not save during initial render
      expect(AsyncStorage.setItem).not.toHaveBeenCalled();
    });

    it('should handle AsyncStorage save errors silently', async () => {
      (AsyncStorage.setItem as jest.Mock).mockRejectedValue(new Error('Save error'));
      
      let cartData: any;
      
      render(
        <CartProvider>
          <TestComponent onRender={(data) => { cartData = data; }} />
        </CartProvider>
      );

      await waitFor(() => {
        expect(cartData).toBeDefined();
      });

      // Should not throw error when save fails
      act(() => {
        cartData.addToCart(mockProduct, 1);
      });

      await waitFor(() => {
        expect(cartData.cart).toHaveLength(1);
      });
    });
  });
});