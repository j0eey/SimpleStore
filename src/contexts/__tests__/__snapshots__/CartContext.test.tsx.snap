// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartContext Cart State Snapshots should match snapshot for cart operations sequence: cart-operations-sequence 1`] = `
[
  {
    "action": "add",
    "cart": [
      {
        "__v": 0,
        "_id": "test-product-1",
        "createdAt": "2024-01-01T00:00:00Z",
        "description": "A test product",
        "images": [
          {
            "_id": "img-1",
            "fullUrl": "https://example.com/test-image.jpg",
            "url": "test-image.jpg",
          },
        ],
        "location": {
          "latitude": 40.7128,
          "longitude": -74.006,
          "name": "Test Location",
        },
        "price": 29.99,
        "quantity": 1,
        "title": "Test Product",
        "updatedAt": "2024-01-01T00:00:00Z",
        "user": {
          "_id": "user-1",
          "email": "test@example.com",
        },
      },
    ],
  },
  {
    "action": "add_existing",
    "cart": [
      {
        "__v": 0,
        "_id": "test-product-1",
        "createdAt": "2024-01-01T00:00:00Z",
        "description": "A test product",
        "images": [
          {
            "_id": "img-1",
            "fullUrl": "https://example.com/test-image.jpg",
            "url": "test-image.jpg",
          },
        ],
        "location": {
          "latitude": 40.7128,
          "longitude": -74.006,
          "name": "Test Location",
        },
        "price": 29.99,
        "quantity": 2,
        "title": "Test Product",
        "updatedAt": "2024-01-01T00:00:00Z",
        "user": {
          "_id": "user-1",
          "email": "test@example.com",
        },
      },
    ],
  },
  {
    "action": "update_quantity",
    "cart": [
      {
        "__v": 0,
        "_id": "test-product-1",
        "createdAt": "2024-01-01T00:00:00Z",
        "description": "A test product",
        "images": [
          {
            "_id": "img-1",
            "fullUrl": "https://example.com/test-image.jpg",
            "url": "test-image.jpg",
          },
        ],
        "location": {
          "latitude": 40.7128,
          "longitude": -74.006,
          "name": "Test Location",
        },
        "price": 29.99,
        "quantity": 5,
        "title": "Test Product",
        "updatedAt": "2024-01-01T00:00:00Z",
        "user": {
          "_id": "user-1",
          "email": "test@example.com",
        },
      },
    ],
  },
  {
    "action": "add_second_product",
    "cart": [
      {
        "__v": 0,
        "_id": "test-product-1",
        "createdAt": "2024-01-01T00:00:00Z",
        "description": "A test product",
        "images": [
          {
            "_id": "img-1",
            "fullUrl": "https://example.com/test-image.jpg",
            "url": "test-image.jpg",
          },
        ],
        "location": {
          "latitude": 40.7128,
          "longitude": -74.006,
          "name": "Test Location",
        },
        "price": 29.99,
        "quantity": 5,
        "title": "Test Product",
        "updatedAt": "2024-01-01T00:00:00Z",
        "user": {
          "_id": "user-1",
          "email": "test@example.com",
        },
      },
      {
        "__v": 0,
        "_id": "test-product-2",
        "createdAt": "2024-01-02T00:00:00Z",
        "description": "Another test product",
        "images": [
          {
            "_id": "img-2",
            "fullUrl": "https://example.com/test-image-2.jpg",
            "url": "test-image-2.jpg",
          },
        ],
        "location": {
          "latitude": 41.8781,
          "longitude": -87.6298,
          "name": "Test Location 2",
        },
        "price": 49.99,
        "quantity": 2,
        "title": "Test Product 2",
        "updatedAt": "2024-01-02T00:00:00Z",
        "user": {
          "_id": "user-2",
          "email": "test2@example.com",
        },
      },
    ],
  },
  {
    "action": "remove_first_product",
    "cart": [
      {
        "__v": 0,
        "_id": "test-product-2",
        "createdAt": "2024-01-02T00:00:00Z",
        "description": "Another test product",
        "images": [
          {
            "_id": "img-2",
            "fullUrl": "https://example.com/test-image-2.jpg",
            "url": "test-image-2.jpg",
          },
        ],
        "location": {
          "latitude": 41.8781,
          "longitude": -87.6298,
          "name": "Test Location 2",
        },
        "price": 49.99,
        "quantity": 2,
        "title": "Test Product 2",
        "updatedAt": "2024-01-02T00:00:00Z",
        "user": {
          "_id": "user-2",
          "email": "test2@example.com",
        },
      },
    ],
  },
]
`;

exports[`CartContext Cart State Snapshots should match snapshot for cart with multiple items: cart-with-multiple-items 1`] = `
{
  "cart": [
    {
      "__v": 0,
      "_id": "test-product-1",
      "createdAt": "2024-01-01T00:00:00Z",
      "description": "A test product",
      "images": [
        {
          "_id": "img-1",
          "fullUrl": "https://example.com/test-image.jpg",
          "url": "test-image.jpg",
        },
      ],
      "location": {
        "latitude": 40.7128,
        "longitude": -74.006,
        "name": "Test Location",
      },
      "price": 29.99,
      "quantity": 2,
      "title": "Test Product",
      "updatedAt": "2024-01-01T00:00:00Z",
      "user": {
        "_id": "user-1",
        "email": "test@example.com",
      },
    },
    {
      "__v": 0,
      "_id": "test-product-2",
      "createdAt": "2024-01-02T00:00:00Z",
      "description": "Another test product",
      "images": [
        {
          "_id": "img-2",
          "fullUrl": "https://example.com/test-image-2.jpg",
          "url": "test-image-2.jpg",
        },
      ],
      "location": {
        "latitude": 41.8781,
        "longitude": -87.6298,
        "name": "Test Location 2",
      },
      "price": 49.99,
      "quantity": 1,
      "title": "Test Product 2",
      "updatedAt": "2024-01-02T00:00:00Z",
      "user": {
        "_id": "user-2",
        "email": "test2@example.com",
      },
    },
  ],
  "cartLength": 2,
  "productIds": [
    "test-product-1",
    "test-product-2",
  ],
  "totalItems": 3,
}
`;

exports[`CartContext Cart State Snapshots should match snapshot for cart with single item: cart-with-single-item 1`] = `
{
  "cart": [
    {
      "__v": 0,
      "_id": "test-product-1",
      "createdAt": "2024-01-01T00:00:00Z",
      "description": "A test product",
      "images": [
        {
          "_id": "img-1",
          "fullUrl": "https://example.com/test-image.jpg",
          "url": "test-image.jpg",
        },
      ],
      "location": {
        "latitude": 40.7128,
        "longitude": -74.006,
        "name": "Test Location",
      },
      "price": 29.99,
      "quantity": 1,
      "title": "Test Product",
      "updatedAt": "2024-01-01T00:00:00Z",
      "user": {
        "_id": "user-1",
        "email": "test@example.com",
      },
    },
  ],
  "cartLength": 1,
  "firstItem": {
    "__v": 0,
    "_id": "test-product-1",
    "createdAt": "2024-01-01T00:00:00Z",
    "description": "A test product",
    "images": [
      {
        "_id": "img-1",
        "fullUrl": "https://example.com/test-image.jpg",
        "url": "test-image.jpg",
      },
    ],
    "location": {
      "latitude": 40.7128,
      "longitude": -74.006,
      "name": "Test Location",
    },
    "price": 29.99,
    "quantity": 1,
    "title": "Test Product",
    "updatedAt": "2024-01-01T00:00:00Z",
    "user": {
      "_id": "user-1",
      "email": "test@example.com",
    },
  },
}
`;

exports[`CartContext Cart State Snapshots should match snapshot for initial empty cart state: initial-empty-cart-state 1`] = `
{
  "availableMethods": [
    "addToCart",
    "removeFromCart",
    "updateCartItemQuantity",
    "clearCart",
  ],
  "cart": [],
  "cartLength": 0,
}
`;

exports[`CartContext Cart State Snapshots should match snapshot for quantity update scenarios: quantity-update-scenarios 1`] = `
[
  {
    "cartState": [
      {
        "__v": 0,
        "_id": "test-product-1",
        "createdAt": "2024-01-01T00:00:00Z",
        "description": "A test product",
        "images": [
          {
            "_id": "img-1",
            "fullUrl": "https://example.com/test-image.jpg",
            "url": "test-image.jpg",
          },
        ],
        "location": {
          "latitude": 40.7128,
          "longitude": -74.006,
          "name": "Test Location",
        },
        "price": 29.99,
        "quantity": 5,
        "title": "Test Product",
        "updatedAt": "2024-01-01T00:00:00Z",
        "user": {
          "_id": "user-1",
          "email": "test@example.com",
        },
      },
    ],
    "newQuantity": 5,
    "scenario": "increase quantity",
  },
  {
    "cartState": [
      {
        "__v": 0,
        "_id": "test-product-1",
        "createdAt": "2024-01-01T00:00:00Z",
        "description": "A test product",
        "images": [
          {
            "_id": "img-1",
            "fullUrl": "https://example.com/test-image.jpg",
            "url": "test-image.jpg",
          },
        ],
        "location": {
          "latitude": 40.7128,
          "longitude": -74.006,
          "name": "Test Location",
        },
        "price": 29.99,
        "quantity": 2,
        "title": "Test Product",
        "updatedAt": "2024-01-01T00:00:00Z",
        "user": {
          "_id": "user-1",
          "email": "test@example.com",
        },
      },
    ],
    "newQuantity": 2,
    "scenario": "decrease quantity",
  },
  {
    "cartState": [
      {
        "__v": 0,
        "_id": "test-product-1",
        "createdAt": "2024-01-01T00:00:00Z",
        "description": "A test product",
        "images": [
          {
            "_id": "img-1",
            "fullUrl": "https://example.com/test-image.jpg",
            "url": "test-image.jpg",
          },
        ],
        "location": {
          "latitude": 40.7128,
          "longitude": -74.006,
          "name": "Test Location",
        },
        "price": 29.99,
        "quantity": 1,
        "title": "Test Product",
        "updatedAt": "2024-01-01T00:00:00Z",
        "user": {
          "_id": "user-1",
          "email": "test@example.com",
        },
      },
    ],
    "newQuantity": 1,
    "scenario": "set to minimum",
  },
  {
    "cartState": [],
    "newQuantity": 0,
    "scenario": "set to zero (should remove)",
  },
]
`;

exports[`CartContext Toast Notification Snapshots should match snapshot for all toast notification calls: toast-notification-calls 1`] = `
[
  [
    {
      "text1": ""Test Product" added to cart!",
      "text2": "Quantity: 1",
      "type": "success",
    },
  ],
  [
    {
      "text1": "Quantity for "Test Product" updated!",
      "text2": "New quantity: 2",
      "type": "info",
    },
  ],
  [
    {
      "text1": ""Test Product 2" added to cart!",
      "text2": "Quantity: 2",
      "type": "success",
    },
  ],
  [
    {
      "text1": "Cart cleared!",
      "type": "info",
    },
  ],
  [
    {
      "text1": "Item removed from cart.",
      "type": "success",
    },
  ],
]
`;
